// @sentry/profiling-node/lib/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.maybeRemoveProfileFromSdkMetadata = exports.isProfiledTransactionEvent = exports.createProfilingEventEnvelope = exports.enrichWithThreadInformation = void 0;
const os_1 = __importDefault(require("os"));
const worker_threads_1 = require("worker_threads");
const utils_1 = require("@sentry/utils");
const THREAD_ID_STRING = String(worker_threads_1.threadId);
const THREAD_NAME = worker_threads_1.isMainThread ? 'main' : 'worker';
// Machine properties (eval only once)
const PLATFORM = os_1.default.platform();
const RELEASE = os_1.default.release();
const VERSION = os_1.default.version();
const TYPE = os_1.default.type();
const MODEL = os_1.default.machine ? os_1.default.machine() : os_1.default.arch();
const ARCH = os_1.default.arch();
function isRawThreadCpuProfile(profile) {
    return !('thread_metadata' in profile);
}
// Enriches the profile with threadId of the current thread.
// This is done in node as we seem to not be able to get the info from C native code.
function enrichWithThreadInformation(profile) {
    if (!isRawThreadCpuProfile(profile)) {
        return profile;
    }
    return {
        samples: profile.samples,
        frames: profile.frames,
        stacks: profile.stacks,
        thread_metadata: {
            [THREAD_ID_STRING]: {
                name: THREAD_NAME
            }
        }
    };
}
exports.enrichWithThreadInformation = enrichWithThreadInformation;
/** Extract sdk info from from the API metadata */
function getSdkMetadataForEnvelopeHeader(metadata) {
    if (!metadata || !metadata.sdk) {
        return undefined;
    }
    return { name: metadata.sdk.name, version: metadata.sdk.version };
}
/**
 * Apply SdkInfo (name, version, packages, integrations) to the corresponding event key.
 * Merge with existing data if any.
 **/
function enhanceEventWithSdkInfo(event, sdkInfo) {
    var _a, _b;
    if (!sdkInfo) {
        return event;
    }
    event.sdk = event.sdk || {};
    event.sdk.name = (_a = (event.sdk.name || sdkInfo.name)) !== null && _a !== void 0 ? _a : 'unknown sdk';
    event.sdk.version = (_b = (event.sdk.version || sdkInfo.version)) !== null && _b !== void 0 ? _b : 'unknown sdk version';
    event.sdk.integrations = [...(event.sdk.integrations || []), ...(sdkInfo.integrations || [])];
    event.sdk.packages = [...(event.sdk.packages || []), ...(sdkInfo.packages || [])];
    return event;
}
function createEventEnvelopeHeaders(event, sdkInfo, tunnel, dsn) {
    const dynamicSamplingContext = event.sdkProcessingMetadata && event.sdkProcessingMetadata['dynamicSamplingContext'];
    return Object.assign(Object.assign(Object.assign({ event_id: event.event_id, sent_at: new Date().toISOString() }, (sdkInfo && { sdk: sdkInfo })), (!!tunnel && { dsn: (0, utils_1.dsnToString)(dsn) })), (event.type === 'transaction' &&
        dynamicSamplingContext && {
        trace: (0, utils_1.dropUndefinedKeys)(Object.assign({}, dynamicSamplingContext))
    }));
}
function createProfilingEventEnvelope(event, dsn, metadata, tunnel) {
    var _a, _b, _c, _d, _e;
    const sdkInfo = getSdkMetadataForEnvelopeHeader(metadata);
    const rawProfile = event.sdkProcessingMetadata['profile'];
    if (event.type !== 'transaction') {
        // createProfilingEventEnvelope should only be called for transactions,
        // we type guard this behavior with isProfiledTransactionEvent.
        throw new TypeError('Profiling events may only be attached to transactions, this should never occur.');
    }
    if (rawProfile === undefined || rawProfile === null) {
        throw new TypeError(`Cannot construct profiling event envelope without a valid profile. Got ${rawProfile} instead.`);
    }
    enhanceEventWithSdkInfo(event, metadata && metadata.sdk);
    const envelopeHeaders = createEventEnvelopeHeaders(event, sdkInfo, tunnel, dsn);
    const enrichedThreadProfile = enrichWithThreadInformation(rawProfile);
    const transactionStartMs = typeof event.start_timestamp === 'number' ? event.start_timestamp * 1000 : Date.now();
    const transactionEndMs = typeof event.timestamp === 'number' ? event.timestamp * 1000 : Date.now();
    const profile = {
        event_id: (0, utils_1.uuid4)(),
        timestamp: new Date(transactionStartMs).toISOString(),
        platform: 'node',
        version: '1',
        release: ((_a = event.sdk) === null || _a === void 0 ? void 0 : _a.version) || 'unknown',
        runtime: {
            name: 'node',
            version: process.versions.node || 'unknown'
        },
        os: {
            name: PLATFORM,
            version: RELEASE,
            build_number: VERSION
        },
        device: {
            locale: (_b = (process.env['LC_ALL'] || process.env['LC_MESSAGES'] || process.env['LANG'] || process.env['LANGUAGE'])) !== null && _b !== void 0 ? _b : 'unknown locale',
            model: MODEL,
            manufacturer: TYPE,
            architecture: ARCH,
            is_emulator: false
        },
        profile: enrichedThreadProfile,
        transactions: [
            {
                name: event.transaction || '',
                id: event.event_id || (0, utils_1.uuid4)(),
                trace_id: (_e = (_d = (_c = event === null || event === void 0 ? void 0 : event.contexts) === null || _c === void 0 ? void 0 : _c['trace']) === null || _d === void 0 ? void 0 : _d['trace_id']) !== null && _e !== void 0 ? _e : '',
                active_thread_id: THREAD_ID_STRING,
                relative_start_ns: '0',
                relative_end_ns: ((transactionEndMs - transactionStartMs) * 1e6).toFixed(0)
            }
        ]
    };
    const envelopeItem = [
        {
            // @ts-expect-error profile is not yet a type in @sentry/types
            type: 'profile'
        },
        // @ts-expect-error profile is not yet a type in @sentry/types
        profile
    ];
    return (0, utils_1.createEnvelope)(envelopeHeaders, [envelopeItem]);
}
exports.createProfilingEventEnvelope = createProfilingEventEnvelope;
function isProfiledTransactionEvent(event) {
    return !!(event.sdkProcessingMetadata && event.sdkProcessingMetadata['profile']);
}
exports.isProfiledTransactionEvent = isProfiledTransactionEvent;
// Due to how profiles are attached to event metadata, we may sometimes want to remove them to ensure
// they are not processed by other Sentry integrations. This can be the case when we cannot construct a valid
// profile from the data we have or some of the mechanisms to send the event (Hub, Transport etc) are not available to us.
function maybeRemoveProfileFromSdkMetadata(event) {
    if (!isProfiledTransactionEvent(event)) {
        return event;
    }
    delete event.sdkProcessingMetadata['profile'];
    return event;
}
exports.maybeRemoveProfileFromSdkMetadata = maybeRemoveProfileFromSdkMetadata;
